[c/en] Wrapping fixes and some copy edits/small additions (#4704)
adambard/learnxinyminutes-docs: c.html.markdown

// libraries for the headers.
// For your own headers, use double quotes instead of angle brackets, and
// provide the path:
#include "my_header.h" // local file
#include "../my_lib/my_lib_header.h" //relative path

// Declare function signatures in advance in a .h file, or at the top of
// your .c file.

> void function_1(void);
> int function_2(int a, float b);

Reviewer: I think it's fine the way it is and is easier to understand when it's simpler like this. I understand `()` is a compiler warning but in practice it doesn't matter if you don't even know how to access the variable length arguments of functions in C. I think mentioning `(void)` at the end is good but keeping it simple at the beginning is better.

```suggestion
void function_1();
int function_2(void);
```

Author: Even with `-Wall -Wextra` gcc (at least version 13.2.0 for me) compiles calling functions with `()` for the parameter list with some arbitrary parameters without any warning. clang does give me some warning. C23 did change this behavior though, making `(void)` and `()` equivalent parameter lists for function declarations, and I know gcc is on v14 now, so my OS's version is probably a bit out of date and this will not become as relevant for new C in not too long.

Having been burning by that unexpected behavior in the past, it was my intuition to include it when making these changes originally, but with the C23 changes putting it off to the end makes sense, or even leaving it out. I'll make some changes shortly

  short x_short = 0;

  // chars are defined as the smallest addressable unit for a processor.

>   // This is usually 1 byte, but for some systems it can be more (ex. for TMS320
>   // from TI it's 2 bytes).

Reviewer: ```suggestion
  // This is usually 1 byte, but for some systems it can be more
  // (ex. for TMS320 from TI it's 2 bytes).
```

  printf("\n");

  // *****NOTES*****:

>   // If the body of control expression is only one statement than parenthesises
>   // are optional, but often recommended:

Reviewer: ```suggestion
  // If the body of control expression is only one statement then parentheses
  // are optional, but recommended:
```

  for (jj = 0; jj < 5; jj++)
      printf("%d\n", jj);

  // Loops and if statements MUST have a body. If no body is needed:
  int i;
  for (i = 0; i <= 5; i++) {
  }
  // Or

>   for (i = 0; i <= 5; i++); // use semicolon to act as the body (null statement)

Reviewer: Moving this comment is wrong, the `;` is a statement terminator on this line, not a body.

Author: Every `for` loop has four parts, one of which is the statement to be executed (aka the body of the for loop). The statement could either be a simple statement, such as a call to `printf` as above, a compound statement (aka block) that is a collection of statements enclosed in `{` and `}`, or the null statement, which is a statement that just doesn't do anything, which is just written as `;`. See [this section of the gnu C manual](https://www.gnu.org/software/gnu-c-manual/gnu-c-manual.html#The-Null-Statement) for reference.

The previous way it was written, it was a null statement inside of a compound statement and the comment implied that the semicolon inside the curlies was necessary when it wasn't. In real life C, for loops that do nothing will have either `;` or `{}`. Although `{;}` as written here would've worked, I've never personally seen it and I don't think it is intuitive.

I certainly think what's going on could be better explained, but I also think that the null statement isn't something that is important enough to spend anything more than a passing mention (if that) on in this introduction to C.

Reviewer: ```suggestion
  for (i = 0; i <= 5; i++); // semicolon acts as the body (null statement)
```

  }
  my_arr[10] = 5;

  // Dereferencing memory that you haven't allocated gives "unpredictable
  // results" - the program is said to invoke "undefined behavior"
  printf("%d\n", *(my_ptr + 21)); // => Prints who-knows-what? It may even crash

  // When you're done with a malloc'd block of memory, you need to free it,

>   // or else no one else can use it until your program terminates (this is
>   // called a "memory leak"):

Reviewer: If it's gonna be on two lines you might as well keep it

```suggestion
  // or else no one else can use it until your program terminates
  // (this is called a "memory leak"):
```

Author: Ifrc correctly lines longer than 80 characters were causing some rendering issues

  free(my_ptr);

  // Strings are arrays of char, but they are usually represented as a
  // pointer-to-char (which is a pointer to the first element of the array).
  // It's good practice to use `const char *' when referring to a string literal

>   // since string literals shall not be modified (i.e. foo[0] = 'a' is ILLEGAL.)

Reviewer: this is definitely clearer

```suggestion
  // since string literals shall not be modified (i.e. "foo"[0] = 'a' is ILLEGAL.)
```

Author: I think this change makes sense but I also think that indexing a string literal will probably make readers take a double take, and I wonder if we could just drop the parenthetical entirely.

// will print "arr[0] is: 1" etc
*/

// if referring to external variables outside function, use the extern keyword.
int i = 0;

> void testFunc(void) {
>   extern int i; //i here is now using external variable i

Reviewer: ```suggestion
void testFunc() {
  extern int i; // i here is now using external variable i
```

}

// make external variables private to source file with static:
static int j = 0; //other files using testFunc2() cannot access variable j

> void testFunc2(void) {

Reviewer: ```suggestion
void testFunc2() {
```

// declared with some other starting value.
//**You may also declare functions as static to make them private**

> // NOTE that before C23, and unlike C++, functions taking no arguments without

Reviewer: ```suggestion
// Note that before C23, and unlike C++, functions taking no arguments without
```

*/

/* A safe guard to prevent the header from being defined too many times. This */
/* happens in the case of circlular dependencies, or such as when a header is */
/* included alongside a header that includes the same header itself.          */
#ifndef EXAMPLE_H /* if EXAMPLE_H is not yet defined. */
#define EXAMPLE_H /* Define the macro EXAMPLE_H. */

> /* This is a less verbose non-standard alternative to the include guards just */
> /* mentioned. While non-standard it avoids errors from the two header files   */
> /* accidentally defining the same macro. Place it at the top of the header.   */
> #pragma once
> 

Reviewer: non standard, shouldn't be mentioned

```suggestion
```
